doc/docgen.rst

```diff
@@ -229,6 +229,146 @@ Output::
+     pi_
+B. qualified (with symbol kind specification)::
+     `const pi`_
+
+For routine kinds there are more options. Consider this definition:
+
+.. code:: Nim
+
+   proc foo*(a: int, b: float): string
+
+Generally following syntax is allowed for referencing `foo`:
+
+*  short (without parameters):
+
+   A. non-qualified::
+
+        foo_
+
+   B. qualified::
+
+        `proc foo`_
+
+*  longer variants (with parameters):
+
+   A. non-qualified:
+
+      1) specifying parameters names::
+
+          `foo(a, b)`_
+
+      2) specifying parameters types::
+
+          `foo(int, float)`_
+
+      3) specifying both names and types::
+
+          `foo(a: int, b: float)`_
+
+      4) output parameter can also be specified if you wish::
+
+          `foo(a: int, b: float): string`_
+
+   B. qualified: all 4 options above are valid.
+      Particularly you can use the full format::
+
+        `proc foo(a: int, b: float): string`_
+
+   **A limitation**:
+   The parameters of a nested routine type can be specified only with types
+   (without parameter names, see form A.2 above). E.g. for this signature:
+
+   .. code:: Nim
+
+      proc binarySearch*[T, K](a: openArray[T]; key: K;
+                               cmp: proc (x: T; y: K): int {.closure.}): int
+                                          ~~    ~~   ~~~~~
+
+   you cannot use names underlined by `~~` so it must be referenced with
+   ``cmp: proc(T, K)``. Hence these forms are valid::
+
+     `binarySearch(a: openArray[T], key: K, cmp: proc(T, K))`_
+     `binarySearch(openArray[T], K, proc(T, K))`_
+     `binarySearch(a, key, cmp)`_
+
+Generic parameters can also be used. All in all, this long form will be
+recognized fine::
+
+  `proc binarySearch*[T, K](a: openArray[T], key: K, cmp: proc(T, K)): int`_
+
+.. Tip:: Avoid extraneous information by using one of these 2 short forms
+   in text::
+
+     binarySearch_
+     `proc binarySearch(a, key, cmp)`_
+
+   If you use a short form and have an ambiguity then just add some
+   additional info.
+
+.. Warning:: An ambiguity in resolving documentation links may arise because of:
+
+   1. clash with other RST anchors
+      * manually setup anchors
+      * automatically set up, e.g. section names
+   2. collision with other Nim symbols.
+      Because in Nim `proc` and `iterator` belong to different namespaces.
+      Note that qualified references are useful in this case: just use the
+      qualified form like `\`proc foo\`_`:literal: or
+      `\`iterator foo\`_`:literal:.
+
+   Any ambiguity is always reported with Nim compiler warnings and an anchor
+   with higher priority is selected. Manual anchors have highest
+   priority, then go automatic RST anchors; then Nim-generated anchors
+   (while procs have higher priority than other Nim symbol kinds).
+
+.. Note:: A bit special case is operators
+   (as their signature is also defined with `\``):
+
+   .. code:: Nim
+
+      func `[]`*[T](x: openArray[T]): T
```

Comment:
should the export marker `*` be present?

Reply:
yes. I mentioned it above:

>This pertains to any exported symbol

Non-exported symbols don't get into documentation (by default, without `--docInternal`) and hence should not be referenced from module description or exported symbols doc comments.

I'm planning to propose a special section `.. internal::` (RST directive) inside doc comments that would be allowed to contain references to non-exported symbols. I expect to send a separate RFC for discussion about literate programming in Nim after finishing this PR and this RFC will contain that feature as its 1st part.

Reply:
sorry ignore my comment, i though you had added the `*` as part of the link